2008-06-13  Ulrich Drepper  <drepper@redhat.com>

        * sysdeps/posix/getaddrinfo.c: Move _res_hconf_init call to a
        better place so it is not called when nscd is used.

2008-05-14  Ulrich Drepper  <drepper@redhat.com>

        * sysdeps/posix/getaddrinfo.c (getaddrinfo): Call _res_hconf_init
        if necessary.
        * posix/tst-rfc3484.c: Add dummy definition of _res_hconf_init.
        * posix/tst-rfc3484-2.c: Likewise.
        * posix/tst-rfc3484-3.c: Likewise.

---
 posix/tst-rfc3484-2.c       |    5 +++++
 posix/tst-rfc3484-3.c       |    6 ++++++
 posix/tst-rfc3484.c         |    5 +++++
 sysdeps/posix/getaddrinfo.c |    4 ++++
 4 files changed, 20 insertions(+)

--- a/posix/tst-rfc3484-2.c
+++ b/posix/tst-rfc3484-2.c
@@ -35,6 +35,11 @@
   *output = NULL;
   return 0;
 }
+void
+attribute_hidden
+_res_hconf_init (void)
+{
+}
 
 #include "../sysdeps/posix/getaddrinfo.c"
 
--- a/posix/tst-rfc3484-3.c
+++ b/posix/tst-rfc3484-3.c
@@ -36,6 +36,12 @@
   return 0;
 }
 
+void
+attribute_hidden
+_res_hconf_init (void)
+{
+}
+
 #include "../sysdeps/posix/getaddrinfo.c"
 
 service_user *__nss_hosts_database attribute_hidden;
--- a/posix/tst-rfc3484.c
+++ b/posix/tst-rfc3484.c
@@ -35,6 +35,11 @@
   *output = NULL;
   return 0;
 }
+void
+attribute_hidden
+_res_hconf_init (void)
+{
+}
 
 #include "../sysdeps/posix/getaddrinfo.c"
 
--- a/sysdeps/posix/getaddrinfo.c
+++ b/sysdeps/posix/getaddrinfo.c
@@ -60,6 +60,7 @@
 #include <not-cancel.h>
 #include <nscd/nscd-client.h>
 #include <nscd/nscd_proto.h>
+#include <resolv/res_hconf.h>
 
 #ifdef HAVE_LIBIDN
 extern int __idna_to_ascii_lz (const char *input, char **output, int flags);
@@ -670,6 +671,9 @@
 					     "dns [!UNAVAIL=return] files",
 					     &nip);
 
+	  /* Initialize configurations.  */
+	  if (__builtin_expect (!_res_hconf.initialized, 0))
+	    _res_hconf_init ();
 	  if (__res_maybe_init (&_res, 0) == -1)
 	    no_more = 1;
 
